docs/authenticate.js

```diff
@@ -195,32 +208,47 @@ function getFirebaseConfig(inProduction) {
 
 /**
  * Initializes Firebase and authenticates using the logged-in Google user
- * (most likely coming from Authenticator above).
+ * coming from Authenticator above.
  *
- * @param {gapi.auth2.AuthResponse} googleAuth
+ * @param {gapi.auth2.GoogleUser} googleUser
  * @return {Promise<any>} Promise that completes when authentication is done
  */
-function authenticateToFirebase(googleAuth) {
+function authenticateToFirebase(googleUser) {
   initializeFirebase();
   return new Promise((resolveFunction) => {
     const unsubscribe = firebase.auth().onAuthStateChanged((firebaseUser) => {
       unsubscribe();
-      if (firebaseUser && firebaseUser.providerData &&
-          firebaseUser.providerData[0].uid) {
-        // The Firebase sample code checks that this is the same user as the
-        // Google user. I don't really see how there can be a mismatch without
-        // something pretty weird going on, so punting on it for now.
-        console.warn('Not logging into Firebase again');
+      if (isUserEqual(googleUser, firebaseUser)) {
         resolveFunction(null);
         return;
       }
       // Build Firebase credential with the Google ID token.
-      const credential =
-          firebase.auth.GoogleAuthProvider.credential(googleAuth.id_token);
+      const credential = firebase.auth.GoogleAuthProvider.credential(
+          googleUser.getAuthResponse().id_token);
       // Sign in with credential from the Google user.
       const signinPromise = firebase.auth().signInWithCredential(credential);
       signinPromise.then(resolveFunction);
       return signinPromise;
     });
   });
 }
+
+/**
+ * Checks if a Firebase user is equal to the given Google user.
+ * @param {gapi.auth2.GoogleUser} googleUser
+ * @param {firebase.User} firebaseUser
+ * @return {boolean}
+ */
+function isUserEqual(googleUser, firebaseUser) {
+  if (firebaseUser) {
```

Comment:
Do we know of a time when this can be fail?

Reply:
I think it might be null the first time you visit the page without having signed in before.